engine/src/test/java/org/hibernate/search/testsupport/TestConstants.java

```diff
@@ -44,43 +43,40 @@ public static Version getTargetLuceneVersion() {
 	}
 
 	/**
-	 * Returns the target directory of the build.
-	 *
-	 * @param testClass the test class for which the target directory is requested.
-	 * @return the target directory of the build
+	 * Returns a temporary directory for storing test data such as indexes etc. Specific tests should store their data
+	 * in sub-directories. The returned directory will be deleted on graceful shut-down. The returned directory will be
+	 * named like {@code $TEMP/hsearch-tests-<random>}.
 	 */
-	public static Path getTargetDir(Class<?> testClass) {
-		ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
-		// get a URL reference to something we now is part of the classpath (our own classes)
-		String currentTestClass = testClass.getName();
-		int hopsToCompileDirectory = currentTestClass.split( "\\." ).length;
-		int hopsToTargetDirectory = hopsToCompileDirectory + 1;
-		URL classURL = contextClassLoader.getResource( currentTestClass.replace( '.', '/' ) + ".class" );
-		// navigate back to '/target'
-		Path targetDir;
+	public static Path getTempTestDataDir() {
 		try {
-			targetDir = Paths.get( classURL.toURI() );
-		}
-		catch (URISyntaxException e) {
-			throw new RuntimeException( e );
+			Path tempTestDataDir = Files.createTempDirectory( "hsearch-tests-" );
+			tempTestDataDir.toFile().deleteOnExit();
+
+			return tempTestDataDir;
 		}
-		// navigate back to '/target'
-		for ( int i = 0; i < hopsToTargetDirectory; i++ ) {
-			targetDir = targetDir.getParent();
+		catch (IOException e) {
+			throw new RuntimeException( "Could not create temporary directory for tests", e );
 		}
-		return targetDir;
 	}
 
 	/**
-	 * Return the root directory to store test indexes in. Tests should never use or delete this directly
-	 * but rather nest sub directories in it to avoid interferences across tests.
+	 * Doing the same as {@link #getIndexDirectory(Path, Class)}, using the immediate caller class as requester.
+	 */
+	public static String getIndexDirectory(Path parent) {
+		return getIndexDirectory( parent, callerProvider.getCallerClass() );
+	}
+
+	/**
+	 * Return the root directory to store test indexes in. Tests should never use or delete this directly but rather
+	 * nest sub directories in it to avoid interferences across tests.
 	 *
-	 * @param testClass the test class for which the index directory is requested.
+	 * @param requester The test class requesting the index directory, its name will be part of the returned directory
 	 * @return Return the root directory to store test indexes
 	 */
-	public static String getIndexDirectory(Class<?> testClass) {
-		Path targetDir = getTargetDir( testClass );
-		Path indexDirPath = targetDir.resolve( "indextemp" );
+	public static String getIndexDirectory(Path parent, Class<?> requester) {
+		Path indexDirPath = parent.resolve( "indextemp-" + requester.getSimpleName() );
+		indexDirPath.toFile().deleteOnExit();
```

Comment:
'deleteOnExit' is just an additional safety right?

I don't think we could rely on it, as I presume it would fail if it's not empty?